src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSProxyContext.cs

```diff
@@ -0,0 +1,601 @@
+        public static JSProxyContext CurrentOperationContext => MainThreadContext;
+
+        public static JSProxyContext PushOperation()
+        {
+            // in single threaded build we don't have to keep stack of operations and the context/thread is always the same
+            return MainThreadContext;
+        }
+#else
+
+        // Context of the main thread
+        private static JSProxyContext? _MainThreadContext;
+        public static JSProxyContext MainThreadContext
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _MainThreadContext!;
+            set => _MainThreadContext = value;
+        }
+
+        private sealed class PendingOperation
+        {
+            public JSProxyContext? CapturedContext;
+            public bool Called;
+            public bool Multiple;
+        }
+
+        [ThreadStatic]
+        private static List<PendingOperation>? _OperationStack;
+        private static List<PendingOperation> OperationStack
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                return _OperationStack ??= new();
+            }
+        }
+
+        public static void PushOperationUnknowContext()
+        {
+            var stack = OperationStack;
+            // for SchedulePopOperation()
+            if (stack.Count > 0 && stack[stack.Count - 1].Called)
+            {
+                PopOperation();
+            }
+            // because this is called multiple times for each JSImport
+            if (stack.Count > 0 && stack[stack.Count - 1].Multiple)
+            {
+                return;
+            }
+
+            stack.Add(new PendingOperation() { Multiple = true });
+        }
+
+        public static void PushOperationWithContext(JSProxyContext knownContext)
+        {
+            var stack = OperationStack;
+            // for SchedulePopOperation()
+            if (stack.Count > 0 && stack[stack.Count - 1].Called)
+            {
+                PopOperation();
+            }
+            // because this is called multiple times for each JSImport
+            if (stack.Count > 0 && stack[stack.Count - 1].Multiple)
+            {
+                return;
+            }
+
+            stack.Add(new PendingOperation() { CapturedContext = knownContext, Multiple = true });
+        }
+
+        public static JSProxyContext PushOperationWithCurrentThreadContext()
+        {
+            var stack = OperationStack;
+            // for SchedulePopOperation()
+            if (stack.Count > 0 && stack[stack.Count - 1].Called)
+            {
+                PopOperation();
+            }
+
+            var current = AssertCurrentContext();
+            stack.Add(new PendingOperation { CapturedContext = current });
+            return current;
+        }
+
+        public static void PopOperation()
+        {
+            var stack = OperationStack;
```

Comment:
we may want an explicit check for an unbalanced stack where we throw an exception with a clear error message